core/src/Cabana_Halo.hpp

```diff
@@ -213,409 +234,670 @@ struct is_halo : public is_halo_impl<typename std::remove_cv<T>::type>::type
+        auto gather_send_buffer_func = KOKKOS_LAMBDA( const std::size_t i )
+        {
+            send_buffer( i ) = aosoa.getTuple( steering( i ) );
+        };
+        Kokkos::RangePolicy<execution_space> gather_send_buffer_policy(
+            0, _halo.totalNumExport() );
+        Kokkos::parallel_for( "Cabana::gather::gather_send_buffer",
+                              gather_send_buffer_policy,
+                              gather_send_buffer_func );
+        Kokkos::fence();
+
+        // The halo has it's own communication space so choose any mpi tag.
+        const int mpi_tag = 2345;
+
+        // Post non-blocking receives.
+        int num_n = _halo.numNeighbor();
+        std::vector<MPI_Request> requests( num_n );
+        std::pair<std::size_t, std::size_t> recv_range = { 0, 0 };
+        for ( int n = 0; n < num_n; ++n )
+        {
+            recv_range.second = recv_range.first + _halo.numImport( n );
+
+            auto recv_subview = Kokkos::subview( recv_buffer, recv_range );
+
+            MPI_Irecv( recv_subview.data(),
+                       recv_subview.size() * sizeof( data_type ), MPI_BYTE,
+                       _halo.neighborRank( n ), mpi_tag, _halo.comm(),
+                       &( requests[n] ) );
+
+            recv_range.first = recv_range.second;
+        }
+
+        // Do blocking sends.
+        std::pair<std::size_t, std::size_t> send_range = { 0, 0 };
+        for ( int n = 0; n < num_n; ++n )
+        {
+            send_range.second = send_range.first + _halo.numExport( n );
+
+            auto send_subview = Kokkos::subview( send_buffer, send_range );
+
+            MPI_Send( send_subview.data(),
+                      send_subview.size() * sizeof( data_type ), MPI_BYTE,
+                      _halo.neighborRank( n ), mpi_tag, _halo.comm() );
+
+            send_range.first = send_range.second;
+        }
+
+        // Wait on non-blocking receives.
+        std::vector<MPI_Status> status( num_n );
+        const int ec =
+            MPI_Waitall( requests.size(), requests.data(), status.data() );
+        if ( MPI_SUCCESS != ec )
+            throw std::logic_error( "Failed MPI Communication" );
+
+        // Extract the receive buffer into the ghosted elements.
+        std::size_t num_local = _halo.numLocal();
+        auto extract_recv_buffer_func = KOKKOS_LAMBDA( const std::size_t i )
+        {
+            std::size_t ghost_idx = i + num_local;
+            aosoa.setTuple( ghost_idx, recv_buffer( i ) );
+        };
+        Kokkos::RangePolicy<execution_space> extract_recv_buffer_policy(
+            0, _halo.totalNumImport() );
+        Kokkos::parallel_for( "Cabana::gather::extract_recv_buffer",
+                              extract_recv_buffer_policy,
+                              extract_recv_buffer_func );
+        Kokkos::fence();
+
+        // Barrier before completing to ensure synchronization.
+        MPI_Barrier( _halo.comm() );
```

Comment:
Would y'all consider getting rid of this `MPI_Barrier`?  You shouldn't actually need it, because this halo exchange has its own communicator (cloned from the user's communicator).  Thus, no other MPI operations will interfere with it, so `MPI_Waitall` suffices to ensure completion.  (If it doesn't, then you have a bug.)

Reply:
Thanks for pointing this out - I'll plan to fix this (and test for bugs) in a follow-up PR